stripecardscan/src/main/java/com/stripe/android/stripecardscan/cardscan/CardScanActivity.kt

```diff
@@ -0,0 +1,571 @@
+        Log.e(Config.logTag, "Canceling scan due to error", cause)
+        runBlocking { scanStat.trackResult("scan_failure") }
+        resultListener.failed(cause)
+        closeScanner()
+    }
+
+    /**
+     * The scan has been closed by the user.
+     */
+    private fun userClosedScanner() {
+        runBlocking { scanStat.trackResult("user_canceled") }
+        resultListener.userCanceled(CancellationReason.Closed)
+        closeScanner()
+    }
+
+    /**
+     * The camera permission was denied.
+     */
+    private fun userDeniedCameraPermission() {
+        runBlocking { scanStat.trackResult("user_canceled") }
+        resultListener.userCanceled(CancellationReason.CameraPermissionDenied)
+        closeScanner()
+    }
+
+    /**
+     * Prepare to start the camera. Once the camera is ready, [onCameraReady] must be called.
+     */
+    private fun prepareCamera(onCameraReady: () -> Unit) {
+        viewBinding.previewFrame.post {
+            viewBinding.viewFinderBackground
+                .setViewFinderRect(viewBinding.viewFinderWindow.asRect())
+            onCameraReady()
+        }
+    }
+
+    private fun onCameraReady() {
+        cameraAdapter.bindToLifecycle(this)
+
+        val torchStat = Stats.trackTask("torch_supported")
+        cameraAdapter.withFlashSupport {
+            launch { torchStat.trackResult(if (it) "supported" else "unsupported") }
+            setFlashlightState(cameraAdapter.isTorchOn())
+            viewBinding.torchButton.setVisible(it)
+        }
+
+        cameraAdapter.withSupportsMultipleCameras {
+            viewBinding.swapCameraButton.setVisible(it)
+        }
+
+        launch { onCameraStreamAvailable(cameraAdapter.getImageStream()) }
+    }
+
+    /**
+     * Turn the flashlight on or off.
+     */
+    private fun setFlashlightState(on: Boolean) {
+        cameraAdapter.setTorchState(on)
+        isFlashlightOn = on
+        if (isFlashlightOn) {
+            viewBinding.torchButton.setDrawable(R.drawable.stripe_flash_on_dark)
+        } else {
+            viewBinding.torchButton.setDrawable(R.drawable.stripe_flash_off_dark)
+        }
+    }
+
+    /**
+     * Set the focus of the camera.
+     */
+    private fun setFocus(point: PointF) {
+        cameraAdapter.setFocus(point)
+    }
+
+    /**
+     * Generate a camera adapter
+     */
+    private fun buildCameraAdapter(): CameraAdapter<CameraPreviewImage<Bitmap>> =
+        getCameraAdapter(
+            activity = this,
+            previewView = viewBinding.previewFrame,
+            minimumResolution = MINIMUM_RESOLUTION,
+            cameraErrorListener = cameraErrorListener,
+        )
+
+    override fun ensureValidParams() = when {
+        params.stripePublishableKey.isEmpty() -> {
+            scanFailure(InvalidStripePublishableKeyException("Missing publishable key"))
+            false
+        }
+        else -> true
+    }
+
+    private fun closeScanner() {
+        setFlashlightState(false)
```

Comment:
We still need to upload scan stats here. ScanStats is how we track the performance of this SDK and determine if anything breaks.